src/bgen/CodeBlocks/CodeBlock.cs

```diff
@@ -0,0 +1,61 @@
+using System.Collections.Generic;
+
+public class CodeBlock : ICodeBlock {
+	protected int CurrentIndent;
+	protected int Indent = 4;
+	protected string HeaderText = string.Empty;
+	protected List<ICodeBlock> Blocks = new ();
+	readonly string startBrace = "{";
+	readonly string endBrace = "}";
+	public readonly string newLine = "\n";
+
+	public CodeBlock (int currentIndent)
+	{
+		this.CurrentIndent = currentIndent;
+	}
+
+	public CodeBlock (int currentIndent, string text)
+	{
+		this.CurrentIndent = currentIndent;
+		HeaderText = text;
+	}
+
+	public void AddBlock (ICodeBlock block)
+	{
+		block.SetIndent (CurrentIndent + Indent);
+		Blocks.Add (block);
+	}
+
+	public void AddLine (string text)
+	{
+		LineBlock line = new LineBlock (CurrentIndent + Indent, text);
+		Blocks.Add (line);
+	}
+
+	protected void DecrementIndent ()
+	{
+		CurrentIndent -= Indent;
+	}
+
+	public virtual string Print ()
```

Comment:
Could Print, rather than return a string work directly on a Stream, being a FileStream of a MemoryStream. It will make the usage more efficient and in reality in the generator we are more interested in streams than in string.